pci/src/msix.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+//
+
+extern crate byteorder;
+extern crate vm_memory;
+
+use crate::{PciCapability, PciCapabilityID};
+use byteorder::{ByteOrder, LittleEndian};
+use vm_memory::ByteValued;
+
+const MAX_MSIX_VECTORS_PER_DEVICE: u16 = 0x800;
+const MSIX_TABLE_ENTRIES_MODULO: u64 = 16;
+const MSIX_PBA_ENTRIES_MODULO: u64 = 8;
+
+#[derive(Debug, Clone)]
+pub struct MsixTableEntry {
+    pub msg_addr_lo: u32,
+    pub msg_addr_hi: u32,
+    pub msg_data: u32,
+    pub vector_ctl: u32,
+}
+
+impl Default for MsixTableEntry {
+    fn default() -> Self {
+        MsixTableEntry {
+            msg_addr_lo: 0,
+            msg_addr_hi: 0,
+            msg_data: 0,
+            vector_ctl: 0,
+        }
+    }
+}
+
+pub struct MsixConfig {
+    pub table_entries: Vec<MsixTableEntry>,
+    pub pba_entries: Vec<u64>,
+}
+
+impl MsixConfig {
+    pub fn new(msix_vectors: u16) -> Self {
+        assert!(msix_vectors < MAX_MSIX_VECTORS_PER_DEVICE);
+
+        let mut table_entries: Vec<MsixTableEntry> = Vec::new();
+        table_entries.resize_with(msix_vectors as usize, Default::default);
+        let mut pba_entries: Vec<u64> = Vec::new();
+        pba_entries.resize_with(msix_vectors as usize, Default::default);
```

Comment:
Since the PBA and the MSI-X entries tables will always be the same length, why not having:

```Rust
pub struct Msix {
    pub msg_addr_lo: u32,
    pub msg_addr_hi: u32,
    pub msg_data: u32,
    pub vector_ctl: u32,
}

pub struct MsixTableEntry {
    pub vector: Msix,
    pub pba: u64,
}

pub struct MsixConfig (pub Vec<MsixTableEntry>);

```

Reply:
They won't. My implementation of PBA is wrong. It's a simple bit array of length "number of vectors". This means in case we have 2 vectors, the PBA will have only 1 entry to cover the 2 bits.
If we had 75 vectors, then we would have 2 entries in the PBA to cover the 75 bits.
I've fixed it now.